File: frontend/src/lib/types/events.ts
import type { InputRequest, PaneInfo, PaneStatus } from './pane';

export interface PaneUpdateEvent {
	event: 'pane_update';
	pane_id: string;
	status: PaneStatus;
	lines: string[];
	input_request?: InputRequest;
	cursor_x: number;
	cursor_y: number;
}

export interface PaneDiscoveryEvent {
	event: 'pane_discovered';
	pane: PaneInfo;
}

export interface PaneRemovedEvent {
	event: 'pane_removed';
	pane_id: string;
}

export interface InitialStateEvent {
	event: 'initial_state';
	panes: PaneInfo[];
}

export interface StateEvent {
	event: 'state';
	panes: PaneInfo[];
}

export type WebSocketEvent =
	| PaneUpdateEvent
	| PaneDiscoveryEvent
	| PaneRemovedEvent
	| InitialStateEvent
	| StateEvent;

File: backend/app/models/events.py
"""WebSocket event models"""
from typing import Literal, Optional, Union

from pydantic import BaseModel, Field

from .input import InputRequest
from .pane import PaneInfo, PaneStatus


class PaneUpdateEvent(BaseModel):
    """Event emitted when pane output or status changes"""
    event: Literal["pane_update"] = "pane_update"
    pane_id: str
    status: PaneStatus
    lines: list[str] = Field(default_factory=list)
    input_request: Optional[InputRequest] = None
    cursor_x: int = 0
    cursor_y: int = 0


class PaneDiscoveryEvent(BaseModel):
    """Event emitted when a new pane is discovered"""
    event: Literal["pane_discovered"] = "pane_discovered"
    pane: PaneInfo


class PaneRemovedEvent(BaseModel):
    """Event emitted when a pane is removed/closed"""
    event: Literal["pane_removed"] = "pane_removed"
    pane_id: str


# Union type for all WebSocket events
WebSocketEvent = Union[PaneUpdateEvent, PaneDiscoveryEvent, PaneRemovedEvent]
